trailbehind/elevation-service
# [GWEB-2033] Elevation Service Refactor

Full refactor of the Elevation Server.

Rewritten in TypeScript, and implements development tooling. 

> The existing tests were simply cloned from the forked repo, and were never updated to work with the changes we made, so they never actually worked. I have written some basic integration tests that _do_ work.

Beyond modernizing the code, the main goal of the refactor was to abstract the concept of "fetch a tile for a [lng,lat], and retrieve some data" into composable and reusable pieces:

1. `fetchTileData` is a general-purpose function that is provided a generic `fetcher` function that does whatever it needs to do to fetch the tile data, and return it in a buffer. There is currently only one fetcher: `s3Fetcher`.
2. The functions in `elevation/` are specific to fetching elevation data, but ultimately use `fetchTileData`.
3. The `server` establishes routes that delegate to functions to handle whatever that route is supposed to do (currently only elevation via `elevation/addElevation.ts`).
4. A "taxonomy" of `Symbol`s is established to coordinate exception handling. Right now, a fetcher can throw `TILE_MISSING`, or `BAD_TILE` as a catch-all (but certainly more specific types could be established if needed). The elevation functions use `NO_DATA` to distinguish that missing elevation data is not an error.

The purpose is to enable use-cases beyond just elevation data, e.g. fetching cellular coverage data.

Follow-up will integrate DataDog so that we can collect critical metrics:
1. Is the LRU cache always at 100% capacity? (probably)
5. How long does it take to fetch from S3 on a cache miss? (seems to occasionally spike)

## Testing

Works locally. Will need to test extensively on staging; thankfully, we now have a staging environment for the elevation service (:high-five: @meaganacombs).

**NOTE**: When running locally, loading tiles from S3 can sometimes take an egregiously long time, like 30s or even a minute. Watch the console output of the elevation server.  Even if the tiles load quickly, it's still likely to take longer than the first retry timeout of 1000ms to load them, so you're likely to see retry messages on the browser console.

See updated README for how to run locally, but it's just:

1. run this with `pnpm dev`
2. run GaiaCloud backend with `pnpm start:backend`
3. run GaiaCloud frontend with `LOCAL_SERVICES=elevation pnpm start:frontend`
